### PR TITLE
Feature; excel; create and update input and output columns

### PR DESCRIPTION
**Pull request by Etschbeijer**
Closes Issue #470 
Enable the addition of new input- and output columns and also update existing ones, when they are already available

### CODE REVIEW

File: src/Client/Update/InterfaceUpdate.fs [hunk 1]
```diff
@@ -102,20 +102,20 @@ module Interface =
...
-            //    model, cmd
+            | Some Swatehost.Excel ->
+                let cmd = OfficeInterop.AddAnnotationBlock minBuildingBlockInfo |> OfficeInteropMsg |> Cmd.ofMsg
+                model, cmd
             | Some Swatehost.Browser | Some Swatehost.ARCitect ->
                 let cmd = Spreadsheet.AddAnnotationBlock minBuildingBlockInfo |> SpreadsheetMsg |> Cmd.ofMsg
                 model, cmd
             | _ -> failwith "not implemented"
-        | AddAnnotationBlocks minBuildingBlockInfos ->
+        | AddAnnotationBlocks compositeColumn ->
```
Comment (Freymaurer) [inline src/Client/Update/InterfaceUpdate.fs:112]: `compositeColumns`

File: src/Client/OfficeInterop/OfficeInterop.fs [hunk 2]
```diff
@@ -798,31 +853,206 @@ let replaceOutputColumn (annotationTableName:string) (existingOutputColumn: Buil
...
 
-            // comment out, to reenable inserting multiple duplicate building block
-            //checkIfBuildingBlockExisting newBB existingBuildingBlocks
+    if inputColumn.IsSome then
 
-            checkHasExistingInput newBB existingBuildingBlocks
-            checkIfBuildingBlockExisting newBB existingBuildingBlocks
-            // if newBB is output column and output column already exists in table this returns (Some outputcolumn-building-block), else None.
-            let outputColOpt = checkHasExistingOutput newBB existingBuildingBlocks
+        let oldInputColumn = annotationTable.columns.items.[(int) inputColumn.Value.index].name
```
Comment (Freymaurer) [inline src/Client/OfficeInterop/OfficeInterop.fs:877]: This is `inputColumn.Value.name`?

File: src/Client/OfficeInterop/OfficeInterop.fs [hunk 3]
```diff
@@ -798,31 +853,206 @@ let replaceOutputColumn (annotationTableName:string) (existingOutputColumn: Buil
...
+let addInputColumn (annotationTable:Table) (arcTable:ArcTable) (newBB:CompositeColumn) =
+
+    if arcTable.TryGetInputColumn().IsSome then
+        failwith "Something went wrong! The add input column is filled with data! Please report this as a bug to the developers."
+
+    else
+
+        let rowCount = arcTable.RowCount + 1
+
+        let createCol name =
```
Comment (Freymaurer) [inline src/Client/OfficeInterop/OfficeInterop.fs:914]: Maybe make this generic and reuseable for other functions

File: src/Client/OfficeInterop/OfficeInterop.fs [hunk 4]
```diff
@@ -798,31 +853,206 @@ let replaceOutputColumn (annotationTableName:string) (existingOutputColumn: Buil
...
+            let! annotationTableName = tryGetActiveAnnotationTableName context
+
+            //When a name is available get the annotation and arctable for easy access of indices and value adaption
+            //Annotation table enables a easy way to adapt the table, updating existing and adding new columns
+            let! annotationTable =
+                if annotationTableName.IsSome then tryGetAnnotationTableByName context annotationTableName.Value
+                else tryGetAnnotationTableByName context ""
+
+            //Arctable enables a fast check for the existence of input- and output-columns and their indices
+            let! arcTable =
```
Comment (Freymaurer) [inline src/Client/OfficeInterop/OfficeInterop.fs:1030]: This should be based on annotationTable and not request from context again.
Then use Option.map for parsing and only ever check if arcTable.IsSome

File: src/Client/OfficeInterop/OfficeInterop.fs [hunk 5]
```diff
@@ -798,31 +853,206 @@ let replaceOutputColumn (annotationTableName:string) (existingOutputColumn: Buil
...
+                else ArcTable.tryFromExcelTableName("", context)
+
+            //When both tables could be accessed succesfully then check what kind of column shall be added an whether it is already there or not
+            if annotationTable.IsSome && arcTable.IsSome then
+                let _ =
+                    annotationTable.Value.rows.load(propertyNames = U2.Case2 (ResizeArray[|"items"; "name"; "values"; "index"; "count"|])) |> ignore
+                    annotationTable.Value.columns.load(propertyNames = U2.Case2 (ResizeArray[|"items"; "name"; "values"; "index"; "count"|]))
+                
+                let! result = context.sync().``then``(fun _ ->
+                    if newBB.Header.isInput then
```
Comment (Freymaurer) [inline src/Client/OfficeInterop/OfficeInterop.fs:1041]: active pattern instead of if-else

**Review by Freymaurer — CHANGES_REQUESTED** [review]

File: src/Client/OfficeInterop/OfficeInterop.fs (hunk 5 above)
Comment (Etschbeijer) [reply src/Client/OfficeInterop/OfficeInterop.fs]: fixed

File: src/Client/OfficeInterop/OfficeInterop.fs (hunk 4 above)
Comment (Etschbeijer) [reply src/Client/OfficeInterop/OfficeInterop.fs:1050]: fixed

File: src/Client/OfficeInterop/OfficeInterop.fs (hunk 3 above)
Comment (Etschbeijer) [reply src/Client/OfficeInterop/OfficeInterop.fs]: fixed

File: src/Client/OfficeInterop/OfficeInterop.fs (hunk 2 above)
Comment (Etschbeijer) [reply src/Client/OfficeInterop/OfficeInterop.fs]: fixed

File: src/Client/Update/InterfaceUpdate.fs (hunk 1 above)
Comment (Etschbeijer) [reply src/Client/Update/InterfaceUpdate.fs]: fixed

File: src/Client/OfficeInterop/OfficeInterop.fs [hunk 6]
```diff
@@ -1024,32 +1044,45 @@ let addAnnotationBlockHandler (newBB:CompositeColumn) =
...
-            if annotationTable.IsSome && arcTable.IsSome then
+            if arcTable.IsSome then
                 let _ =
                     annotationTable.Value.rows.load(propertyNames = U2.Case2 (ResizeArray[|"items"; "name"; "values"; "index"; "count"|])) |> ignore
                     annotationTable.Value.columns.load(propertyNames = U2.Case2 (ResizeArray[|"items"; "name"; "values"; "index"; "count"|]))
-                
-                let! result = context.sync().``then``(fun _ ->
-                    if newBB.Header.isInput then
+
+                let (|Input|_|) (newBuildingBlock:CompositeColumn) =
```
[the remark this answers is not in the document]
Comment (Etschbeijer) [reply src/Client/OfficeInterop/OfficeInterop.fs:1062]: Ok, das soll ich dann nicht in diesem Issue, anschließen, richtig?
Das würde ich auf später verschieben, sobald alle einzelnen Building Blocks fertig sind, dann habe ich die Übersicht (und vermutlich Erfahrung), das generische Konstrukt aufzuziehen, ok?

File: src/Client/OfficeInterop/OfficeInterop.fs [hunk 7]
```diff
@@ -51,6 +51,14 @@ module OfficeInteropExtensions =
             |> Seq.map (snd >> Seq.map snd >> Seq.toArray)
             |> Seq.toArray
 
+        static member addCollumn (annotationTable:Table) name rowCount value =
```
[the remark this answers is not in the document]
Comment (Etschbeijer) [reply src/Client/OfficeInterop/OfficeInterop.fs]: fixed

File: src/Client/OfficeInterop/OfficeInterop.fs [hunk 8]
```diff
@@ -798,31 +893,199 @@ let replaceOutputColumn (annotationTableName:string) (existingOutputColumn: Buil
...
+/// <summary>
+/// Update an existing inputcolumn of an annotation table
+/// </summary>
+/// <param name="excelTable"></param>
+/// <param name="arcTable"></param>
+/// <param name="newBB"></param>
+let updateInputColumn (excelTable:Table) (arcTable:ArcTable) (newBB:CompositeColumn) =
+
+    let possibleInputColumn = arcTable.TryGetInputColumn()
+    let inputColumnName =
```
Comment (Freymaurer) [inline src/Client/OfficeInterop/OfficeInterop.fs:905]: Do not default to column name to "" if it is empty, just work with options. No reason to run the code below if we get a None here.

```fsharp
Option.map (fun c -> c.Header.ToString())
|> Option.map (fun cName -> excelTable.columns.item.Find (fun s -> ...)
```

Comment (Freymaurer) [reply src/Client/OfficeInterop/OfficeInterop.fs:905]: Same for output

File: src/Client/OfficeInterop/OfficeInterop.fs [hunk 9]
```diff
@@ -798,31 +893,199 @@ let replaceOutputColumn (annotationTableName:string) (existingOutputColumn: Buil
...
+let addAnnotationBlockHandler (newBB:CompositeColumn) =
     Excel.run(fun context ->
         promise {
 
-            let! annotationTableName = getActiveAnnotationTableName context
-            let sheet = context.workbook.worksheets.getActiveWorksheet()
-            let annotationTable = sheet.tables.getItem(annotationTableName)
+            //let activeSheet = context.workbook.worksheets.getActiveWorksheet()
+            //Try to get the name of the currently active sheet
+            let! excelTableName = tryGetActiveAnnotationTableName context
```
Comment (Freymaurer) [inline src/Client/OfficeInterop/OfficeInterop.fs:1041]: fail if this isNone, then shadow it with its own .Value

File: src/Client/OfficeInterop/OfficeInterop.fs (hunk 6 above)
[the remark this answers is not in the document]
Comment (Freymaurer) [reply src/Client/OfficeInterop/OfficeInterop.fs:1062]: As you see fit. But the current implementation of `let (|Input|_|) (newBuildingBlock:CompositeColumn) =`, `let (|Output|_|) (newBuildingBlock:CompositeColumn) =` seems more complicated than it helps.

**Review by Freymaurer — CHANGES_REQUESTED** [review]

File: src/Client/OfficeInterop/OfficeInterop.fs (hunk 8 above)
Comment (Etschbeijer) [reply src/Client/OfficeInterop/OfficeInterop.fs]: fixed

File: src/Client/OfficeInterop/OfficeInterop.fs (hunk 9 above)
Comment (Etschbeijer) [reply src/Client/OfficeInterop/OfficeInterop.fs:1048]: fixed

**Review by Freymaurer — APPROVED** [review]

**Pull request merged**